MergerLogic/DataTypes/S3.cs

```diff
@@ -1,136 +1,173 @@
-                this._endOfRead = !response.IsTruncated;
-            }
-
-            return tiles;
-        }
-
-        public override void setBatchIdentifier(string batchIdentifier)
-        {
-            this._continuationToken = batchIdentifier;
-        }
-
-        public override bool Exists()
-        {
-            var listRequests = new ListObjectsV2Request
-            {
-                BucketName = this._bucket,
-                Prefix = this.Path,
-                StartAfter = this.Path,
-                MaxKeys = 1
-            };
-
-            this._logger.LogInformation($"Checking if exists, bucket: {this._bucket}, path: {this.Path}");
-            var task = this._client.ListObjectsV2Async(listRequests);
-            var response = task.Result;
-            return response.KeyCount > 0;
-        }
-
-        public override long TileCount()
-        {
-            long tileCount = 0;
-            string continuationToken = null;
-
-            do
-            {
-                var listRequests = new ListObjectsV2Request
-                {
-                    BucketName = this._bucket,
-                    Prefix = this.Path,
-                    StartAfter = this.Path,
-                    ContinuationToken = continuationToken
-                };
-
-                var task = this._client.ListObjectsV2Async(listRequests);
-                var response = task.Result;
-
-                tileCount += response.KeyCount;
-                continuationToken = response.NextContinuationToken;
-            } while (continuationToken != null);
-
-            return tileCount;
-        }
-
-        protected override void InternalUpdateTiles(IEnumerable<Tile> targetTiles)
-        {
-            foreach (var tile in targetTiles)
-            {
-                this.Utils.UpdateTile(tile);
-            }
-        }
-    }
-}
+using Amazon.S3;
+using Amazon.S3.Model;
+using MergerLogic.Batching;
+using MergerLogic.Clients;
+using MergerLogic.Utils;
+using Microsoft.Extensions.Logging;
+
+namespace MergerLogic.DataTypes
+{
+    public class S3 : Data<IS3Client>
+    {
+        private readonly IAmazonS3 _client;
+        private readonly string _bucket;
+        private IEnumerator<int> _zoomLevels;
+        private string? _continuationToken;
+        private bool _endOfRead;
+
+        private readonly IPathUtils _pathUtils;
+
+        public S3(IPathUtils pathUtils, IAmazonS3 client, IServiceProvider container,
+            string bucket, string path, int batchSize, Grid? grid, GridOrigin? origin)
+            : base(container, DataType.S3, path, batchSize, grid, origin)
+        {
+            this._pathUtils = pathUtils;
+            this._bucket = bucket;
+            this._continuationToken = null;
+            this._endOfRead = false;
+            this._client = client;
+
+            // This should always happen after the definition of the client
+            this._zoomLevels = this.GetZoomLevels();

```

Comment:
why not converting it to array and then no reset is needed? its not enough data to be memory problem